### DISCUSSION TITLE
Dory Discards Issue

### DISCUSSION THREAD

**Discussion by 20Rishabh20** [dspeterson/dory#27] [General]
Getting a lot of dory discards with error logs:-
1) Discarding message due to rate limit
2) Discarding message because failed delivery attempt limit reached
3) Discarding message due to buffer space cap

Current Configuration is:-
<?xml version="1.0" encoding="US-ASCII"?>

<doryConfig>
    <batching>
        <namedConfigs>
            <config name="low_latency">
                

                
                <time value="500" />

                
                <messages value="disable" />

                
                <bytes value="256k" />
            </config>

            <config name="default_latency">
                <time value="10000" />
                <messages value="disable" />
                <bytes value="256k" />
            </config>
        </namedConfigs>

        
        <produceRequestDataLimit value="1024k" />

        
        <messageMaxBytes value="1000000" />

        
        <combinedTopics enable="true" config="default_latency" />

        
        <defaultTopic action="combinedTopics" config="" />

        <topicConfigs>
            
        </topicConfigs>
    </batching>

    <compression>
        <namedConfigs>
            
            <config name="snappy_config" type="snappy" minSize="128" />

            
            <config name="no_compression" type="none" />
        </namedConfigs>

        
        <sizeThresholdPercent value="75" />

        
        <defaultTopic config="snappy_config" />

        <topicConfigs>
            
        </topicConfigs>
    </compression>

    <topicRateLimiting>
        <namedConfigs>
            
            <config name="zero" interval="1" maxCount="0" />

            
            <config name="infinity" interval="1" maxCount="unlimited" />

            
            <config name="config1" interval="10000" maxCount="1000" />

            
            <config name="config2" interval="15000" maxCount="4k" />
        </namedConfigs>

        
        <defaultTopic config="infinity" />

        <topicConfigs>
            
            <topic name="topic1" config="zero" />
            <topic name="topic2" config="infinity" />
            <topic name="topic3" config="config1" />
            <topic name="topic4" config="config2" />
        </topicConfigs>
    </topicRateLimiting>

    <initialBrokers>
        
	<broker host="ts005i" port="9092" />

	<broker host="ts008i" port="9092" />
    </initialBrokers>
</doryConfig>

Current Dory Version:- 2.0.5
OS version:- centos 6.10
Please help with the issue.

**Comment by dspeterson** [comment]
Hi 20Rishabh20,

I'm currently on vacation and won't have easy access to a computer until next weekend, so until then my ability to provide help is limited.  The per-topic message rate limiting feature was implemented quite a while ago, and the intent was to prevent buggy clients from overwhelming the Kafka brokers with too many messages.  I believe recent versions of Kafka have pretty good built-in throttling functionality to deal with this sort of problem, so the rate limiting feature is probably not as useful as it once it was.  To stop the rate limit discards, you can loosen or eliminate the rate limits in the config file.  For instance, to eliminate them completely, you can edit the topic configs under the <topicRateLimiting> section to look like this:
```
    <topicConfigs>
    </topicConfigs>
```
or like this:
```
    <topicConfigs>
        <!-- Rate limit configurations for individual topics go here. -->
        <topic name="topic1" config="infinity" />
        <topic name="topic2" config="infinity" />
        <topic name="topic3" config="infinity" />
        <topic name="topic4" config="infinity" />
    </topicConfigs>
```
The discards due to failed delivery attempt limit reached and buffer space cap indicate some issue interacting with the Kafka brokers.  Does Dory's logfile have any error messages that may indicate why the failed delivery attempts are occurring?  You can post log excerpts here and I'll take a look.  Another useful diagnostic tool is the counters Dory provides through its web interface as documented [here](https://github.com/dspeterson/dory/blob/master/doc/status_monitoring.md).  The failed delivery attempts should cause one or more counters to increment, indicating why they are occurring.  You can match up the counter increments with specific places in the code where the counters are being incremented to get more information on what's going on.  If you want, you can post counter output here and I'll take a look.  I hope this helps!

Dave
